text/trace/0205-messaging-semantic-conventions-context-propagation.md

```diff
@@ -0,0 +1,177 @@
+If the message creation context cannot be attached to the message and
+propagated, consumer traces cannot be directly correlated to producer traces.
+
+### Requirements
+
+A producer SHOULD attach a message creation context to each message. The message creation context
+SHOULD be attached in a way so that it is not possible to be changed by intermediaries.
+
+## Future possibilities
+
+### Transport context propagation
+
+A message creation context can be attached to a message, while different
+contexts are propagated with requests that publish and fetch a message. When
+coming up with conventions and guidance for intermediary instrumentation, it
+will be beneficial to clearly outline those two layers of context propagation
+and build conventions for intermediary instrumentation on top of this outline:
+
+1. The _message context layer_ allows correlating the producer with the
+   consumers of a message, regardless of intermediary instrumentation. The
+   creation context is created by the producer and must be propagated to the
+   consumers. It must not be altered by intermediaries.
+
+   This layer helps to model dependencies between producers and consumers,
+   regardless of the underlying messaging transport mechanism and its
+   instrumentation.
+2. An additional _transport context layer_ allows correlating the producer and
+   the consumer with an intermediary. It also allows correlating multiple
+   intermediaries among each other. The transport context can be changed by
+   intermediaries, according to intermediary instrumentations. Intermediaries that
+   are not instrumented might simply drop the transport context.
+
+   This layer helps to gain insights into details of the message transport.
+
+This would keep the existing correlation between producers and consumers intact
+while allowing intermediaries to use the transport context to correlate
+intermediary instrumentation with existing producer and consumer
+instrumentations.
+
+### Standards for context propagation
+
+Currently, instrumentation authors have to decide how to attach and extract
+context from messages to fulfil the [requirements for context propagation](#context-propagation).
+While preserving the freedom for instrumentation authors to choose how to
+propagate context, in the future these conventions should list recommended ways
+of how to propagate context using well-established messaging protocols.
+
+There are several work-in-progress efforts to standardize context propagation for different
+messaging protocols and scenarios:
+
+* [AMQP](https://w3c.github.io/trace-context-amqp/)
+* [MQTT](https://w3c.github.io/trace-context-mqtt/)
+* [CloudEvents via HTTP](https://github.com/cloudevents/spec/blob/v1.0.1/extensions/distributed-tracing.md)
+
+Once the standards reach a stable state, these semantic conventions will give a
+clear and stable recommendation for each protocol and scenario.
```

Comment:
I think we want the standards to reach a stable state **and** define how both message creation context and transport context should be carried in the message. The goal is to provide direction here what we would like the standards to achieve before being adopted by these semantic conventions.

Here's an example of what I think could go wrong without specific direction. The current MQTT draft standard defines a single context only. One path forward could be for the MQTT standard to declare that "this context is the message creation context and must not be modified by intermediaries" in the draft. This would allow it to be adopted here, but this would leave a gap for transport context propagation. 

Even if these conventions don't define transport layer tracing for messaging now, we want it to evolve to include transport-layer tracing. So I think we should provide guidance as to what is generally needed by those standards for messaging.

```suggestion
Once the standards reach a stable state and define how the message creation
context and transport context are represented, these semantic conventions will
give a clear and stable recommendation for each protocol and scenario.
```

Reply:
That makes sense to me. If nobody objects, I will accept this change.